src/ethers/signer.ts

```diff
@@ -57,43 +68,119 @@ export class KlaytnWallet extends Wallet {
   async populateTransaction(transaction: Deferrable<TransactionRequest>): Promise<TransactionRequest> {
     let tx: TransactionRequest = await resolveProperties(transaction);
 
-    if (!TypedTxFactory.has(tx.type)) {
+    if (!KlaytnTxFactory.has(tx.type)) {
       return super.populateTransaction(tx);
     }
 
+    if ( !(tx.gasPrice) ) {
+      if (this.provider instanceof JsonRpcProvider ) {
+        const result = await this.provider.send("klay_gasPrice", []);
+        tx.gasPrice = result;
+      } else {
+        throw new Error(`Klaytn transaction can only be populated from a Klaytn JSON-RPC server`);
+      }
+    }
+
+    if ( !(tx.gasLimit) && !!(tx.to) ) {
+      if (this.provider instanceof JsonRpcProvider ) {
+
+        const estimateGasAllowedKeys: string[] = [
+          "from", "to", "gasLimit", "gasPrice", "value", "input" ];
+        let ttx = encodeTxForRPC( estimateGasAllowedKeys, tx );
+
+        const result = await this.provider.send("klay_estimateGas", [ttx]);
+        tx.gasLimit = result*2;  // multiply 2 for enuough gas
```

Comment:
Please look into ethers.js and Metamask, and how they calculate the gasLimit from eth_estimateGas result.
If possible, add the GitHub link for the examples.

Reply:
In ethers.js, they just call eth_estimateGas without multiplying like us. 

https://github.com/ethers-io/ethers.js/blob/c785c1e5153d5a85a78f90f177577f2c8df15529/src.ts/providers/abstract-signer.ts#L80
https://github.com/ethers-io/ethers.js/blob/c785c1e5153d5a85a78f90f177577f2c8df15529/src.ts/providers/provider-jsonrpc.ts#L511 
https://github.com/ethers-io/ethers.js/blob/c785c1e5153d5a85a78f90f177577f2c8df15529/src.ts/providers/provider-jsonrpc.ts#L764

Reply:
In Metamask, multiply 1.5 is used. 
And it also compares blockGasLimit. 
It will be nice to have simply the same factor 1.5. 

https://github.com/MetaMask/metamask-extension/blob/9d38e537fca4a61643743f6bf3409f20189eb8bb/ui/ducks/send/helpers.js#L115
https://github.com/MetaMask/metamask-extension/blob/9d38e537fca4a61643743f6bf3409f20189eb8bb/ui/pages/send/send.utils.js#L46

Reply:
If that's the case, let's follow ethers.

If SDK users (wallet or dapp devs) want to add some margin, they can always
- Manually call `eth_estimateGas` or `klay_estimateGas` and multiply by a factor.
- Edit the populatedTx (`tx.gas = tx.gas * 1.5`)

Users indifferent to gas shall face out-of-gas errors. They should
- Raise the gasLimit to large number
- Read the example and use enough number
- Learn how Klaytn Tx intrinsic gas are calculated - which is unlikely because there's no documentation for it. You should see the source code for the info (e.g. VTwithMemo intrinsic gas is `21000 + len(memo)*100` - https://github.com/klaytn/klaytn/blob/dev/blockchain/types/tx_internal_data_value_transfer_memo.go#L239)

Reply:
I updated a guideline to the example by referencing your comment.

https://github.com/blukat29/klaytn-sdk-design/pull/2/commits/14b7d75c4af8569b3be90a416a6c005a98ef9ece

Reply:
I make sure there is no logic to modify specially like ethers.js,
and modified the role for adjusting for inaccurate results of estimateGas is in dapp or wallet like Metamask.

https://github.com/blukat29/klaytn-sdk-design/pull/2/commits/a4c1e67591d00212c8a54d75a0c249fe84663cdd